napari/conftest.py

```diff
@@ -470,3 +472,166 @@ def _mock_app():
             yield app
         finally:
             Application.destroy('test_app')
+
+
+def _get_calling_place(depth=1):
+    if hasattr(sys, "_getframe"):
+        frame = sys._getframe(1 + depth)
+        return f"{frame.f_code.co_filename}:{frame.f_code.co_firstlineno}"
+    return ""
+
+
+@pytest.fixture
+def dangling_threads(monkeypatch):
+    from qtpy.QtCore import QThread
+
+    base_start = QThread.start
+    thread_dkt = WeakKeyDictionary()
+
+    def my_start(self, priority=QThread.InheritPriority):
+        thread_dkt[self] = _get_calling_place()
+        base_start(self, priority)
+
+    monkeypatch.setattr(QThread, 'start', my_start)
+    yield
+
+    dangling_threads = []
+
+    for thread, calling in thread_dkt.items():
+        try:
+            if thread.isRunning():
+                dangling_threads.append((thread, calling))
+        except RuntimeError as e:
+            if "wrapped C/C++ object of type QThread" not in e.args[0]:
+                raise
+
+    for thread, _ in dangling_threads:
+        thread.quit()
+        thread.wait(2000)
+
+    assert not dangling_threads, "thread called in:\n" + "\n".join(
+        x[1] for x in dangling_threads
+    )
+
+
+@pytest.fixture
+def dangling_thread_pool(monkeypatch):
+    from qtpy.QtCore import QThreadPool
+
+    base_start = QThreadPool.start
+    threadpool_dkt = WeakKeyDictionary()
+
+    def my_start(self, runnable, priority=0):
+        if self not in threadpool_dkt:
+            threadpool_dkt[self] = []
+        threadpool_dkt[self].append(_get_calling_place())
+        base_start(self, runnable, priority)
+
+    monkeypatch.setattr(QThreadPool, 'start', my_start)
+    yield
+
+    dangling_threads_pools = []
+
+    for thread_pool, calling in threadpool_dkt.items():
+        if thread_pool.activeThreadCount():
+            dangling_threads_pools.append((thread_pool, calling))
+
+    for thread_pool, _ in dangling_threads_pools:
+        thread_pool.clear()
+        thread_pool.waitForDone(2000)
+
+    assert not dangling_threads_pools, "thread called in:\n" + "\n".join(
+        "; ".join(x[1]) for x in dangling_threads_pools
+    )
+
+
+@pytest.fixture
+def dangling_timers(monkeypatch):
+    from qtpy.QtCore import QTimer
+
+    base_start = QTimer.start
+    timer_dkt = WeakKeyDictionary()
+    single_shot_list = []
+
+    def my_start(self, msec=None):
+        timer_dkt[self] = _get_calling_place()
+        if msec is not None:
+            base_start(self, msec)
+        else:
+            base_start(self)
+
+    monkeypatch.setattr(QTimer, 'start', my_start)
+
+    def single_shot(msec, reciver, method=None):
+        t = QTimer()
+        t.setSingleShot(True)
+        if method is None:
+            t.timeout.connect(reciver)
+        else:
+            t.timeout.connect(getattr(reciver, method))
+        single_shot_list.append((t, _get_calling_place(2)))
+        base_start(t, msec)
```

Comment:
Is this intended to cover all the [`QTimer::singleShot` static convenience methods right](https://doc.qt.io/qt-5/qtimer.html#static-public-members)? Seems like there a few missing. Or are some of these dropped in pyqt/pyside?

Reply:
It looks like i need to sit and fix it